### ISSUE TITLE
`dzil build` should empty the target directory when starting up

### ISSUE THREAD

**Issue by karenetheridge** [rjbs/Dist-Zilla#521]
as noted here -- https://rt.cpan.org/Ticket/Display.html?id=111984#txn-1596865 -- if the build directory already exists, `dzil build` will add its files to the existing contents rather than emptying it out. This can yield undesired results if the build configuration was changed, as files can be left over from last time that didn't get overwritten.  The directory should probably be emptied out at the start of the build.

**Comment by rjbs** [comment]
…and probably prompt before doing so.  Sounds good.

**Comment by tlhackque** [comment]
I don't object to a prompt - but there should be a way to suppress it.

Usually the build directories are a scratch area that can be iterated on many times with trial runs & packaging experiments.  And people shouldn't be doing work inside them.  (though some probably do.)

So while a prompt isn't a bad thing - it could, for example, catch a case where one forgets to update a version number - it should be possible to avoid the prompt.

Perhaps a dist.ini options such as PromptIfBuildDirectoryDirty = yes|no.

I won't quibble about the default.

**Comment by rjbs** [comment]
That all sounds fines to me except for that option name.  Good grief, man! 😄

**Comment by karenetheridge** [comment]
It could be a command-line option on the `build` command? (and `test`, `release`, `smoke` too, I guess...), to avoid promping? I guess I don't mind which way it defaults as I use a shell alias to do a build anyway.

**Comment by karenetheridge** [comment]
For the `release` command especially I think it's important to empty out the directory first -- otherwise the uploaded distribution isn't exactly what was intended to be built.

**Comment by tlhackque** [comment]
Besides shipping what you intend, you'd really like your tests prior to release to run in the same environment as release.

To be clear, if you say 'no' to the prompt, dz should abort the build.

A build should **always** empty the directory.  The only point to a prompt is to allow for _oops, for some reason the wrong directory is being used._  E.g. the build directory name comes from the version, which sometimes people forget to update.  

A dz build proves that your kit is complete.  If it relies on left-over junk, it's a false proof.  If it ships stuff that found its way into the build directory by some other path, the shipped kit is incorrect.  Maybe mildly so.  Maybe some confidential material gets packaged...

**Any build in a non-empty directory is defective**.  The only choices are to change to another directory, or empty it.

**Comment by rjbs** [comment]
Yes, agreed: the point of the prompt is not to allow you to not empty the dir and still use.  It's to allow you to abort and fix the dir you're using, if needed.

**Comment by dolmen** [comment]
About prompting, see #514.

**Comment by karenetheridge** [comment]
After re-reading all of this, I think a patch should be pretty
straight-forward.  At the start of the build, when we figure out the directory
name to build into, we check that it either 1. doesn't exist, or 2. exists but
is totally empty. If that check fails, we either kill the build right now, or
prompt the user to consent to emptying out the dir first before proceeding --
which of these depends on the value of option passed to the build/release
command. (the test command isn't affected because it always creates a brand
new temp dir for the build.)

All that we really need is some bikeshedding on the option name, and what the
default value should be.  Please advise?

I thought of `--force`, which mirrors the -f option for mv/cp/rm ("I don't
care if the destination already exists - overwrite it if you have to"), but
that might be a little too vague. any other ideas?

**Comment by karenetheridge** [comment]
summary:

    $ dzil build
    [DZ] that build dir already exists. ok to delete? [Y/n]
    Y  -> dir gets deleted, then build continues
    N  -> [DZ] aborting build

    $ dzil build --force
    [DZ] <build dir> already exists; deleting...
    [DZ] ..build continues...

**Comment by tlhackque** [comment]
That seems fine to me.  Thanks for reviving this.

On the default: usually we make defaults non-destructive, which argues for 'N'.  The argument for 'Y" would be that a build directory is temporary and that Y is the expected answer.  I'm OK with either, with a slight preference for Y.  I expect to alias dzbuild="dzil build --force"...

If the build is run non-interactively (see IO::Interactive), it should fail with a message & non-zero status but without prompting.  (e.g. as if --noforce was specified).  

If you don't like --force, consider --clean (From the usual Makefile target).

**Comment by karenetheridge** [comment]
Defaulting to "die, rather than deleting something that the user might have wanted to save" sounds fine to me.

Dist::Zilla prompts (see Dist::Zilla::Chrome::Term) all revert to picking the default if not interactive, so defaulting to dying would be safest in that case too.

**Comment by karenetheridge** [comment]
FWIW, my opening comment here is totally wrong:

> as noted here -- https://rt.cpan.org/Ticket/Display.html?id=111984#txn-1596865 -- if the build directory already exists, dzil build will add its files to the existing contents rather than emptying it out. This can yield undesired results if the build configuration was changed, as files can be left over from last time that didn't get overwritten. The directory should probably be emptied out at the start of the build.

We can see by inspecting Dist::Zilla::Dist::Builder::_prep_build_root that a
`remove_tree` is performed on the build directory.

The proposed changes in this issue would simply give us the chance to preserve
files in the build directory before they are deleted, rather than prevent
leakage between builds.  It's still a worthwhile thing to have, but quite a
different problem than the one I described.

**Comment by tlhackque** [comment]
In the original RT, I reported an observation that the build behavior changed depending on whether or not the directory exists at the start of a build.  There was a MANIFEST mismatch error UNLESS the build directory existed from a previous run.  IIRC, it had to do with the build script creating files - a changelog from git and README from POD.  Apparently, the MANIFEST enumerator doesn't (or didn't) recognize that a RunBefore could create files.  The gory details escape me, as this was over a year ago.  It was reproducible at that time.

However, the build behavior should not depend on whether the directory exists.  So perhaps the `remove_tree` that you found doesn't run early enough?  If it did, I would have expected to see the MANIFEST mismatch whether or not the build directory existed (and/or was empty),

I've long since worked around the issue in my builds, and backtracking to recreate the issue isn't worthwhile for me.

The suggestion that there should be an option to preserve the contents of an existing directory didn't come from me (see the first reply in this thread).  I'm OK if it exists - but don't view it as essential.

If dzil currently unconditionally empties the build directory before running any part of the build, I don't see the need for yet another option.  

It might be worth checking to see whether anything (including RunBefores) happens before the remove_tree that you found is executed....

**Comment by karenetheridge** [comment]
The ordering is thus:  most of the build occurs (file gathering, file munging, prereqs injected, ...) in memory, then the build directory is prepared (created or emptied out as necessary), then the files are written out, then the AfterBuild plugins run. If we are releasing rather than building, then the BeforeRelease, Release and AfterRelease plugins are run next.

Whether or not the build directory already exists prior to files being written out *shouldn't* matter, because none of the plugins have any access to the name of that build directory. They shouldn't be able to do anything with it because they should have no idea what it is. The only files on disk that should be mattering to the build are those that are in the 'source' directory (which might be the current working directory, but it also might not be).

> Apparently, the MANIFEST enumerator doesn't (or didn't) recognize that a RunBefore could create files

As I stated at the time in that ticket, any files you want to be added to the distribution need to be added in memory to the zilla object (the infrastructure for this is provided through Dist::Zilla::Role::FileGatherer), *not* created as a physical file.  If you had a `[Run::BeforeBuild]` plugin that was attempting to guess the eventual target build directory and create files in there, then that itself is a bug.  It sounds like you've solved it already though, so yay!

**Comment by brunoramoslu** [comment]
I just got bitten by this "behaviour" by calling:

`dzil build --indir ..`

It just deleted the whole parent directly without prompting.

I see that there is already a special case to manage the situation where the dot would be passed to dzil build

```
$ carton exec "dzil -v build --in ."
using "--in ." would destroy your working directory!
```

I do think this needs to be implemented to stop any unintended deletes.
I'll take a look at the code to try to produce a patch.
